openedx_authz/api/data.py

```diff
@@ -0,0 +1,227 @@
+        """The username for the user (e.g., 'john_doe').
+
+        This is an alias for external_key that represents the username without the namespace prefix.
+
+        Returns:
+            str: The username without namespace.
+        """
+        return self.external_key
+
+
+@define
+class ActionData(AuthZData):
+    """An action is an operation that can be performed in a specific scope.
+
+    Attributes:
+        action: The action name. Automatically prefixed with 'act@' if not present.
+    """
+
+    NAMESPACE: str = "act"
+    name: str = ""
+
+    @property
+    def name(self) -> str:
+        """The human-readable name of the action (e.g., 'Delete Library', 'Edit Content').
+
+        This property transforms the external_key into a human-readable display name
+        by replacing underscores with spaces and capitalizing each word.
+
+        Returns:
+            str: The human-readable action name (e.g., 'Delete Library').
+        """
+        return self.external_key.replace("_", " ").title()
+
+
+@define
+class PermissionData(AuthZData):
+    """A permission is an action that can be performed under certain conditions.
+
+    Attributes:
+        name: The name of the permission.
+    """
+
+    action: ActionData = None
+    effect: Literal["allow", "deny"] = "allow"
+
+
+@define
+class RoleMetadataData(AuthZData):
+    """Metadata for a role.
+
+    Attributes:
+        description: A description of the role.
+        created_at: The date and time the role was created.
+        created_by: The ID of the subject who created the role.
+    """
+
+    description: str = None
+    created_at: str = None
+    created_by: str = None
+
+
+@define
+class RoleData(AuthZData):
+    """A role is a named group of permissions.
+
+    Attributes:
+        name: The name of the role. Must have 'role@' namespace prefix.
+        role_id: The role identifier namespaced (e.g., 'role@instructor').
+        permissions: A list of permissions assigned to the role.
+        metadata: A dictionary of metadata assigned to the role. This can include
+            information such as the description of the role, creation date, etc.
+    """
+
+    NAMESPACE: str = "role"
+    permissions: list[PermissionData] = None
+    metadata: RoleMetadataData = None
+
+    @property
+    def name(self) -> str:
+        """The human-readable name of the role (e.g., 'Library Admin', 'Course Instructor').
+
+        This property transforms the external_key into a human-readable display name
+        by replacing underscores with spaces and capitalizing each word.
+
+        Returns:
+            str: The human-readable role name (e.g., 'Library Admin').
+        """
+        return self.external_key.replace("_", " ").title()
+
+
+@define
+class RoleAssignmentData(AuthZData):
+    """A role assignment is the assignment of a role to a subject in a specific scope.
+
+    Attributes:
+        subject: The ID of the user namespaced (e.g., 'user@john_doe').
```

Comment:
There is where I think it gets weird to use `@` as the separator

Reply:
Yes! I changed the implementation so it's easier to change the namespace/separators and also changed the default separator to ^: https://github.com/openedx/openedx-authz/pull/75/commits/ccc108817a53f7ba2a774c99f28bea83c75aa50a